Allow admin user to explicitly unschedule the node

pravisankar opened the pull request:
Node deactivation will block scheduling of new pods on the node.

Commit:
Allow admin user to explicitly unschedule the node

Setting Unschedulable on the node will not touch any existing pods
on the node but will block scheduling of new pods on the node.

pravisankar commented:
Related discussion on https://github.com/GoogleCloudPlatform/kubernetes/issues/3885
@bgrant0607 @ddysher @smarterclayton

pravisankar commented:
Covered under Red Hat corporate CLA

ddysher commented:
We shouldn't have the custom logic in apiserver.  Also, as @bgrant0607 mentioned, status should be reconstructable and be observed by k8s components, we can't have user update status I think.

An alternative proposal is to update node spec, have node controller check the spec and update status accordingly.

bgrant0607 commented:
Discussed w/ @ddysher today.

There should be 2 fields in NodeSpec: Schedulable and Runnable. These reflect disabling of scheduling new pods and execution of any user pods, respectively, by administrative action, independent of unplanned failures.

There should also be 2 new NodeConditions, Schedulable and Runnable. The spec fields and other conditions (e.g., Ready) would be combined to provide the status values. The scheduler would then only need to look at the Schedulable condition, in the normal case.

davidopp commented:
> There should be 2 fields in NodeSpec: Schedulable and Runnable. These reflect disabling of 
> scheduling new pods and execution of any user pods, respectively, by administrative action, 
> independent of unplanned failures.

Should these be label selectors rather than binary values? Internally have a number of use cases where the administrator wants to make some subset of pods un-{schedulable,runnable} as opposed to it being a binary decision.

> There should also be 2 new NodeConditions, Schedulable and Runnable

I'm not completely opposed to this, but it seems a little weird to put the logic for determining whether a node is schedulable into a controller rather than in the scheduler. Runnable seems less weird since node controller is already making decisions based on health checks.

bgrant0607 commented:
Yes, this model was deliberately simpler than we know it will eventually need to be. We don't have any notion of priority yet, nor forgiveness, nor disruption SLOs, nor pre-stop hook timeouts (graceful termination deadlines). 

I also thought about making nodes fit into the model for graceful termination of all objects: #1535. That's intended for preparation for object deletion, which wouldn't normally be the case for maintenance, but we could do something similar.

We could add a stop time goal, expected duration, and reason to the Spec, and no new conditions, and leave it entirely to the scheduler to assess schedulability and to the controller to determine runnability for any particular pod. That would provide enough information for forgiveness, disruption SLOs, and respecting graceful termination expectations. The scenarios where one wants to selectively disable certain classes of workloads in the cluster or on individual machines for other reasons are more rare.

However, I'm concerned that a more sophisticated model would be harder for users to understand, and we don't yet have any of the features that would make it necessary.

If we wanted the scheduler to only need to look at the node's status and not the spec, we could simply reflect the administrative bits into the conditions, without combining with other conditions, which is what I had previously proposed in the original NodeCondition PR.

ddysher commented:
@pravisankar Some summary and backgroud:  First, we need two new specs: Runnable and Schedulable.  We can start with simple bool, but eventually, we add more information to support pod forgiveness (#1574), etc.  Second, we have three conditions: Runnable, Schedulable and Ready (we'll remove Reachable for now).  Kubelet will periodically push Ready status; node controller will periodically check spec and Ready status to populate other two conditions. At the moment, Node controller will also do pod eviction if Ready status is not updated for a long time.  Third, scheduler will only schedule pod to schedulable nodes.  As a precaution, it's up to scheduler to check other condition too. 

As @bgrant0607 mentioned, there is a lot policies ommitted, but let's make it fancier later.  The change involves kubelet, scheduler and node controller.  I'll start kubelet POST status shortly, let me know what I can help at your side.

@davidopp  runnable and schedulable is a notion of a node, by "make subset of pods un-schedulable/runnable", I guess you mean make a node unschedulable for certain type of pods?  I don't think we have enough features to do this, even if we do, I think we should also use nodespec.  Label is for indetifying information, whereas schedulable/runnable is status (reflection of specification). Label can be anything, but we need a clear spec for the model.

davidopp commented on docs/node.md:
````diff
@@ -125,3 +131,9 @@ A Kubernetes administrator typically uses `kubectl` to manage `Node`. Similar
+will not affect any existing pods on the node but it will disable creation of
+any new pods on the node. Node deactivate example:
+```
+kubectl update nodes 10.1.2.3 --patch='{"apiVersion": "v1beta1", "deactivate": true}'
````
since it only blocks scheduling, maybe call it "stopScheduling" or something like that? deactivate seems ambiguous, and might be interpreted as also killing the running pods

davidopp commented on pkg/api/types.go:
```diff
@@ -776,8 +776,11 @@ type EndpointsList struct {
 	Capacity ResourceList `json:"capacity,omitempty"`
+
+	// Deactivate controls node schedulability of new pods. By default node is schedulable.
+	Deactivate bool `json:"deactivate,omitempty"`
```
As mentioned above, I would suggest something like "Unschedulable" rather than "Deactivate"

bgrant0607 replied:
+1

pravisankar commented:
As per the feedback,
Renamed node spec field 'deactivate' to 'unschedule'.
Moved the node unschedule logic from apiserver to node controller.
Runnable condition will be introduced in a seperate pull request.
@ddysher please review the changes

davidopp commented on pkg/api/types.go:
```diff
@@ -776,8 +776,11 @@ type EndpointsList struct {
 	Capacity ResourceList `json:"capacity,omitempty"`
+
+	// Unschedule controls node schedulability of new pods. By default node is schedulable.
+	Unschedule bool `json:"unschedule,omitempty"`
```
Please call this MakeUnschedulable.

davidopp commented on api/swagger-spec/v1beta1.json:
```diff
@@ -3308,6 +3308,10 @@
       "type": "v1beta1.NodeResources",
       "description": "characterization of node resources"
      },
+     "unschedule": {
```
makeUnschdulable (and similar changes in the other files)

pravisankar commented:
@davidopp @ddysher renamed 'unschedule' to 'makeUnschedulable'.

bgrant0607 commented:
Sorry to give contradictory advice, but the spec is declarative intent -- the desired state. The field names should not sound like imperative commands. It should be `Unschedulable`.

ddysher commented:
+1 that we should use `unschedulable`.  Thanks for the change, I'll take a look later today.

pravisankar commented:
@bgrant0607 @ddysher Okay, renamed to `unschedulable`
On the same lines (declarative intent), don't we want to rename `CreateExternalLoadBalancer` to `ExternalLoadBalancer` ?

ddysher commented on docs/node.md:
```diff
@@ -125,3 +131,9 @@ A Kubernetes administrator typically uses `kubectl` to manage `Node`. Similar
 to Node Controller, `kubectl` command only creates/deletes node representation.
 Note if Kubernetes is running on cloud provider, `kubectl create` a node will
 be refused if Node Controller has already synchronized nodes from cloud provider.
+Admin can choose to unschedule the node using `kubectl`. Unscheduling the node
```
s/unschedule the node/make the node unschedulable

ddysher commented on docs/node.md:
```diff
@@ -38,13 +38,15 @@ must have appropriate conditions, see below.
+conditions are `NodeReachable`, `NodeReady` and `NodeSchedulable`. In the
+future, we plan to add more. `NodeReachable` means the node can be reached
+within the cluster. `NodeReady` means the kubelet returns StatusOK for HTTP
+health check. `NodeSchedulable` means node is ready to schedulable any new
```
s/schedulable/schedule

ddysher commented on pkg/cloudprovider/controller/nodecontroller.go:
```diff
@@ -292,12 +285,48 @@ func (s *NodeController) DoCheck(node *api.Node) []api.NodeCondition {
 
+	// Check Condition: NodeSchedulable
+	oldSchedulableCondition := s.getCondition(node, api.NodeSchedulable)
+	newSchedulableCondition := s.checkNodeSchedulable(node)
```
Do we want to set node to unschedulable if node is not ready?

pravisankar replied:
I think node status should reflect the current status and it may have multiple node conditions in failed state. Subset of the node conditions may impact other components in the system.

ddysher replied:
What does it mean if a node is schedulable but not ready?  A schedulable node should be ready. We are supporting plugin scheduler, so we need to make sure each condition is sematically right.

bgrant0607 replied:
Our choice is to combine conditions in the scheduler, here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L182

Or to combine them here. 

Given that I foresee handling eviction in the node controller, I agree that it makes sense to do it here.

ddysher commented on pkg/cloudprovider/controller/nodecontroller_test.go:
```diff
@@ -537,10 +537,16 @@ func TestHealthCheckNode(t *testing.T) {
 		{
-			node: newNode("node0"),
+			node: &api.Node{ObjectMeta: api.ObjectMeta{Name: "node0"},
+				Spec: api.NodeSpec{Unschedulable: false}},
```
nit: do not add new line for a single struct.

ddysher commented on pkg/cloudprovider/controller/nodecontroller.go:
```diff
@@ -292,12 +285,48 @@ func (s *NodeController) DoCheck(node *api.Node) []api.NodeCondition {
+		return &api.NodeCondition{
+			Type:          api.NodeSchedulable,
+			Status:        api.ConditionNone,
+			Reason:        "User specified during node create/update.",
```
nit: everywhere in our code, Reason doesn't end with '.'

bgrant0607 replied:
Reason is intended to be a single word -- everywhere Reason exists in the API.  Use Message for a sentence.

ddysher commented:
The interaction between node controller and other components becomes significant now.  I think we'll need to add e2e test / integration test later.

I can speak to servicespec, but that does sound like an inconsistency.  `ExternalLoadBalanced` sounds better to me.

pravisankar commented:
@ddysher updated

bgrant0607 commented:
Good point about createExternalLoadBalancer. Will file an issue.

pravisankar commented:
@ddysher @bgrant0607 
I was under the impression that NodeScedulable is a condition controlled by user and it doesn't worry about other conditions (like NodeReady) and scheduler will inspect all the conditions to deduce whether to include/exclude the node during decision making.
Now I understand that we have inherent ordering in the conditions, NodeSchedulable = true/full implies node has to be reachable and ready and NodeSchedulable = false means either user has explicitly stated not to be part of pod scheduling or node is not reachable/ready.
Fixed this issue and corresponding test cases. please review the changes.

davidopp commented:
I think I finally understand why the logic works the way it does. Correct me if I'm wrong. The reasoning is basically
* In general, Spec is input to a component, Status is output from the component. [ Maybe we can document this pattern in one of our development or architecture guides? ] Component we're concerned about here is NodeController.
* Here, Unschedulable is input to NodeController (e.g. from client), so it's part of NodeSpec
* Here, NodeSchedulable is output from NodeController (and input to scheduler), so it's part of NodeStatus (via NodeCondition)

I think this part is fine. The thing I'm not so sure about is whether we should have the middle of the three conditions in checkNodeSchedulable() -- the one that sets schedulable to false if the node is not reachable. (Basically, the stuff you wrote about in your last comment.) In theory, someone might want to write a scheduler that still considers a node schedulable even if it is not reachable. The way the code is currently structured, they'd need to write a new controller and a new scheduler if they wanted to implement this policy. If we provide schedulable and reachable as separate inputs to the scheduler (and don't have reachable control schedulable) then the scheduler can set whatever policy it wants for combining these inputs.

pravisankar commented:
@davidopp 
Good point, I updated  checkNodeSchedulable() based on these comments:

<snip>
 > +	newSchedulableCondition := s.checkNodeSchedulable(node)

What does it mean if a node is schedulable but not ready? A schedulable node should be ready. We are supporting plugin scheduler, so we need to make sure each condition is semantically right.

Our choice is to combine conditions in the scheduler, here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L182
Or to combine them here.
Given that I foresee handling eviction in the node controller, I agree that it makes sense to do it here.
</snip>

Contemplating more on this, now I seem to like what I did in the first pass i.e. NodeSchedulable condition not considering other conditions in the node controller. `kubectl get nodes` returning  node status as 'Schedulable,NotReady,Reachable' is more informative than  'NotSchedulable,NotReady,Reachable'. I think figuring out node is schedulable or not has to be the job of the scheduler by inspecting all conditions on the node.

NodeSchedulable=true when NodeReady=false may look symantically incorrect. May be we just need a better name for this condition,  `NodeMarkedSchedulable` instead of `NodeSchedulable`. This removes any symantic confusion when NodeMarkedSchedulabe=true and NodeReady=false.
As David pointed out, this also eliminates writing any new controller if some scheduler wants to consider node schedulable condition even if it is not reachable.

@davidopp @bgrant0607 @ddysher

davidopp commented:
> I think figuring out node is schedulable or not has to be the job of the scheduler by inspecting all conditions on the node.

Yes, exactly. Combining them in the node controller loses information. There's no good reason to do that, and it limits flexibility in policies other components can implement. Changing the name to NodeMarkedSchedulable does seem helpful if you change the implementation to the approach you mentioned.

I don't think the approach you use here has any bearing on handling eviction in the node controller later, because schedulability has nothing to do with eviction anyway. 

I agree with Brian that "schedulable but not ready" doesn't seem like a useful combination right now, but allowing them to be independent conditions gives more flexibility to the scheduler and thus I think ti is a net win.

ddysher commented:
For the moment, either way is ok.  Maybe use something like `schedulability` instead of `NodeMarkedSchedulable` (`NodeMarkedSchedulable` sounds a bit verbose).

In the long run, we'll have a more complex model, who will be responsible to decide if a node is schedulable?  E.g. if we add a type `NodeMaintenanceEvent`, I expect node controller to update node status accordingly, and scheduler to decide how to schedule a node based on the status.  This is really a joint effort from both controller and scheduler.  Right now, the schedulable is only a boolean, so it can be put at either side, i.e. at controller or scheduler.

davidopp commented:
@pravisankar please update the schedulable status logic as we've discussed and then I think this is fine to merge (after one last review).

pravisankar commented:
@davidopp @ddysher updated as per above discussion.
I didn't rename `NodeSchedulable` (if we use `NodeSchedulability` then we need to use `NodeReachability` to make it consistent). If you strongly feel we need to rename, please let me know and I'll be happy to make the change.

ddysher commented on docs/node.md:
```diff
@@ -38,13 +38,15 @@ must have appropriate conditions, see below.
+health. Kubernetes will make a comprehensive scheduling decision based on the
+information. Node condition is represented as a json object. For example, the
+following conditions mean the node is reachable from its cluster, node is in
+sane state but not ready to accept pods:
```
Isn't {Schedulable: Full, Ready:None} the conditions for what your are describing here?

pravisankar replied:
No
Ready:Full ==> Node is healthy (sane state)
Schedulable: None ==> Node marked as unscedulable (not ready to accept pods)

ddysher replied:
This confuses people as it says "not ready".  It also doesn't highlight the fact that schedulable is controlled via spec.

ddysher commented on api/swagger-spec/v1beta1.json:
```diff
@@ -4342,6 +4342,10 @@
       "type": "uint64",
       "description": "string that identifies the internal version of this object that can be used by clients to determine when objects have changed; value must be treated as opaque by clients and passed unmodified back to the server"
      },
+     "unschedulable": {
```
I'd be surprised if we have to add them manually....

pravisankar replied:
Is there a script that will auto generate these values from api/types.go ?

bgrant0607 replied:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/update-swagger-spec.sh

ddysher replied:
Have you ran the script to verify the spec?

ddysher commented on pkg/cloudprovider/controller/nodecontroller_test.go:
```diff
@@ -552,7 +552,7 @@ func TestSyncCloudDeletePods(t *testing.T) {
 }
 
-func TestHealthCheckNode(t *testing.T) {
+func TestNodeConditionsCheck(t *testing.T) {
```
As this is expanding quickly, can you add some comments on each case:
1. test setup
2. expectation

We shouldn't test by method, we need to test by behavior, i.e. this method probably needs to be splitted to multiple ones.  I'm fine with it right now.

ddysher commented on pkg/cloudprovider/controller/nodecontroller_test.go:
```diff
@@ -704,6 +729,12 @@ func TestSyncNodeStatusTransitionTime(t *testing.T) {
 								},
+								{
+									Type:               api.NodeSchedulable,
+									Status:             api.ConditionFull,
```
Shouldn't this be ConditionNone?  Though I think probably doesn't matter for this method (Test Transitiontime).

pravisankar replied:
No, Check comment in L690, existing node is schedulable so it should be ConditionFull.

ddysher replied:
I meant to comment on next one...  github's review folds too much.  But looks like you also comment existing node is also schedulable.  It looks weired though, with spec not schedulable and condition scheduable

pravisankar replied:
The use case we are testing here is to verify last transition time change for the node condition. For this scenario, test case setup:
(a) Current node status is schedulable 
(b) Now user desired node to be not schedulable (updated in spec) but status is not updated yet  to reflect this desire.
Test execution:
   Test case will update the status (obeys the spec and marks NodeSchedulable to ConditionNone) and ensures there is change in last transition time.

ddysher commented on pkg/cloudprovider/controller/nodecontroller.go:
```diff
@@ -343,12 +336,48 @@ func (s *NodeController) DoCheck(node *api.Node) []api.NodeCondition {
+		return &api.NodeCondition{
+			Type:          api.NodeSchedulable,
+			Status:        api.ConditionNone,
+			Reason:        "User specified during node create/update",
```
The reason message can't stand on its own, maybe "User marked unschedulable during node create/update"?

pravisankar commented:
@ddysher @davidopp updated, please review

bgrant0607 commented:
If spec vs. status is not clear, we should clarify api-conventions.md.

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -704,6 +706,8 @@ type Minion struct {
 	// Pod IP range assigned to the node
 	PodCIDR string `json:"cidr,omitempty" description:"IP range assigned to the node"`
+	// Unschedulable controls node schedulability of new pods. By default node is schedulable.
+	Unschedulable bool `json:"unschedulable,omitempty" description:"enable or disable pod scheduling on the node"`
```
Actually, true => disable, so I'd omit "enable or".

bgrant0607 commented on pkg/cloudprovider/controller/nodecontroller.go:
```diff
@@ -343,12 +336,48 @@ func (s *NodeController) DoCheck(node *api.Node) []api.NodeCondition {
+}
+
+// checkNodeSchedulable checks node schedulable condition, without transition timestamp set.
+func (s *NodeController) checkNodeSchedulable(node *api.Node) *api.NodeCondition {
```
I'm fine with this behavior.

ddysher commented:
@pravisankar Thanks!  Few minor comments.

pravisankar commented:
@ddysher @davidopp updated

erictune commented:
Needs rebase.

pravisankar commented:
rebased PTAL

ddysher commented:
Needs rebase. LGTM

pravisankar commented:
rebased and updated swagger spec using hack/update-swagger-spec.sh

ddysher commented:
/cc @nikhiljindal Can you take a look at the swagger spec?

bgrant0607 commented:
@ddysher Unfortunately, the swagger contents are rearranged every time that script is run, so it's hard to inspect carefully. Hopefully we'll fix that soon. For now, we should just merge if we're happy with the changes.

bgrant0607 commented:
Unfortunately it needs to be rebased again, though, and probably needs to re-update the swagger.

Commit:
Updated swagger spec

pravisankar commented:
rebased (updated swagger spec one more time)

ddysher commented:
LGTM, thanks for your work!
